mava/systems/tf/madqn/training.py

```diff
@@ -593,65 +626,131 @@ def _forward(self, inputs: Any) -> None:
         Args:
             inputs (Any): input data from the data table (transitions)
         """
+        # Get info about the samples from reverb.
+        sample_info = inputs.info
+        sample_keys = tf.transpose(inputs.info.key)
+        sample_probs = tf.transpose(sample_info.probability)
+
+        # Initialize sample priorities at zero.
+        sample_priorities = np.zeros(len(inputs.info.key))
 
         data = tree.map_structure(
             lambda v: tf.expand_dims(v, axis=0) if len(v.shape) <= 1 else v, inputs.data
         )
         data = tf2_utils.batch_to_sequence(data)
 
-        observations, actions, rewards, discounts, _, _ = (
-            data.observations,
-            data.actions,
-            data.rewards,
-            data.discounts,
-            data.start_of_episode,
-            data.extras,
-        )
+        observations, actions, rewards, discounts, _, extras = data
 
-        # Using extra directly from inputs due to shape.
-        core_state = tree.map_structure(
+        # core_states = extra["core_states"]
+        core_states = tree.map_structure(
             lambda s: s[:, 0, :], inputs.data.extras["core_states"]
         )
 
+        # Initial states.
+        states = {agent: core_states[agent][0] for agent in self._agents}
+        target_states = {agent: core_states[agent][0] for agent in self._agents}
+
+        q_network_losses: Dict[str, NestedArray] = {}
         with tf.GradientTape(persistent=True) as tape:
-            q_network_losses: Dict[str, NestedArray] = {}
 
+            # Unroll over time dimension.
+            T = list(observations.values())[0].observation.shape[0]  # time dimension
+            q_stacks: Dict[str, list] = {agent: [] for agent in self._agents}
+            q_target_stacks: Dict[str, list] = {agent: [] for agent in self._agents}
+            for t in range(T):
+
+                # Step each agent.
+                for agent in self._agents:
```

Comment:
Why is two loops over the agents needed for recurrent DQN instead of one loop (e.g. MADDPG)?

Reply:
It is because I want to build communication directly into here instead of having a separate trainer for it and when doing communication, at every time step each agent sends a message.

Reply:
With regard to experiments on Flatland: I am setting things up on GCP now. Will hopefully have some results in a couple hours.